src/sync/sync_session.cpp

```diff
@@ -461,44 +454,35 @@ bool SyncSession::can_wait_for_network_completion() const
     return m_state == &State::active || m_state == &State::dying;
 }
 
-void SyncSession::wait_for_upload_completion()
+bool SyncSession::wait_for_upload_completion(std::function<void(std::error_code)> callback)
 {
     std::unique_lock<std::mutex> lock(m_state_mutex);
     if (can_wait_for_network_completion()) {
         REALM_ASSERT(m_session);
-        m_session->wait_for_upload_complete_or_client_stopped();
+        m_session->async_wait_for_upload_completion(std::move(callback));
+        return true;
     }
+    return false;
 }
 
-void SyncSession::wait_for_download_completion()
+bool SyncSession::wait_for_download_completion(std::function<void(std::error_code)> callback)
 {
     std::unique_lock<std::mutex> lock(m_state_mutex);
     if (can_wait_for_network_completion()) {
         REALM_ASSERT(m_session);
-        m_session->wait_for_download_complete_or_client_stopped();
+        m_session->async_wait_for_download_completion(std::move(callback));
+        return true;
     }
+    return false;
 }
 
-void SyncSession::wait_for_upload_completion(std::function<void()> callback)
-{
-    // FIXME: If the session is waiting for a token, the `wait_for_upload` should be deferred until the `bind()`
-    // instead of just calling the callback immediately.
-    REALM_ASSERT(shared_from_this());
-    std::thread([callback=std::move(callback), self=shared_from_this()]() {
-        self->wait_for_upload_completion();
-        callback();
-    }).detach();
-}
-
-void SyncSession::wait_for_download_completion(std::function<void()> callback)
+void SyncSession::wait_for_upload_completion_blocking()
 {
-    // FIXME: If the session is waiting for a token, the `wait_for_upload` should be deferred until the `bind()`
```

Comment:
The tests that were depending on that behavior (in Cocoa) no longer use this API. That being said, it's probably a good idea to ensure the async API actually respects this behavior (i.e. doesn't call the callback immediately if `bind()` hasn't yet been called).